u5-rest-apis/features/book-controller.js

```diff
@@ -0,0 +1,251 @@
+  try {
+    // 1. Inputs
+    const isbn13 = req.params.isbn13
+
+    // 2. Query
+    const query = db.collection('books').doc(isbn13).delete()
+
+    // 3. Response
+    await query
+    res.sendStatus(200)
+  } catch (err) {
+    console.error(err)
+    res.sendStatus(500)
+  }
+}
+
+/// ///////////////////////////// Covid \\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\
+
+const readCovidRecords = async (req, res) => {
+  try {
+    // 1. Inputs
+    // none
+
+    // 2. Query
+    const query = db.collection('covid-latest').get()
+
+    // 3. Response
+    const payload = (await query)
+      .docs
+      .map(doc => doc.data())
+      .map(({ date, stateId, stateName, cases, casesNew, vaccineOne, vaccineOnePercent, vaccineComplete, vaccineCompletePercent }) =>
+        ({ date: date.toMillis(), stateId, stateName, cases, casesNew, vaccineOne, vaccineOnePercent, vaccineComplete, vaccineCompletePercent }))
+
+    res.json({
+      result: 'ok',
+      payload: payload,
+      count: payload.length
+    })
+  } catch (err) {
+    console.error(err)
+    res.status(500).json({
+      result: 'error',
+      payload: [],
+      count: 0
+    })
+  }
+}
+
+const readCovidRecord = async (req, res) => {
+  try {
+    // 1. Inputs
+    const stateId = req.params.stateId.toUpperCase()
+    // 2. Query
+    const query = db.collection('covid-history')
+      .doc(stateId)
+      .get()
+
+    // 3. Response
+    const snapshot = await query
+    if (!snapshot.exists) { return res.status(404).json({ result: 'not found' }) }
+
+    const { stateName, history } = snapshot.data()
+    const payload = { stateId, stateName, history: history.map(item => ({ date: item.date.toMillis(), cases: item.cases, casesNew: item.casesNew, vaccineOne: item.vaccineOne, vaccineOnePercent: item.vaccineOnePercent, vaccineComplete: item.vaccineComplete, vaccineCompletePercent: item.vaccineCompletePercent })) }
+    res.json({ result: 'ok', payload })
+  } catch (err) {
+    console.error(err)
+    res.status(500).json({ result: 'error' })
+  }
+}
+
+const createCovidRecord = async (req, res) => {
+  try {
+    // 1. Inputs
+    const { date, stateId, cases, casesNew, vaccineOne, vaccineOnePercent, vaccineComplete, vaccineCompletePercent } = req.body
+    const record = { stateId, date: firestore.Timestamp.fromMillis(date), cases, casesNew, vaccineOne, vaccineOnePercent, vaccineComplete, vaccineCompletePercent }
+
+    // 2. Query
+    const query = db.collection('covid-latest')
+      .doc(stateId)
+      .set(record, { merge: true })
+
+    // 3. Response
+    await query
+    res.sendStatus(201)
+  } catch (err) {
+    console.error(err)
+    res.sendStatus(500)
+  }
+}
+
+const UpdateCovidRecord = async (req, res) => {
+  try {
+    // 1. Inputs
+    const stateId = req.params.stateId
+    const { casesNew } = req.body
+    const record = { casesNew }
+
+    // 2. Query
+    const query = db.collection('covid-latest').doc(stateId).set(record, { merge: true })
```

Comment:
หมายเหตุ: สำหรับห้องปฏิบัติการ คุณควรถือว่า `date` ในคำขอเท่ากับ `date` ใน `covid-latest` (นี่คือการอัปเดตข้อมูล "ล่าสุด") โซลูชันที่สมจริงยิ่งขึ้นจะตรวจสอบว่าสิ่งนี้เป็นจริงโดยใช้ธุรกรรม -- ตัวอย่างเช่น การอัปเดตข้อมูลสำหรับวันที่เก่าไม่ควรเปลี่ยน `covid-latest` มันจะอัพเดทแค่ `covid-history`

Note: For the lab you should assume that `date` is equal to the `date` already in `covid-latest` (this is updating the "latest" data). A more realistic solution would check that this is true using a transaction -- for example updating data for an old date should not change `covid-latest`; it would only update `covid-history`.